pkg/endpointmanager/manager.go

```diff
@@ -75,6 +86,98 @@ func NewEndpointManager(epSynchronizer EndpointResourceSynchronizer) *EndpointMa
+	Check(*endpoint.Endpoint) error
+	DeleteEndpoint(*endpoint.Endpoint) int
+}
+
+// WithPeriodicEndpointGC runs a controller to periodically garbage collect
+// endpoints that match the specified checker.
+func (mgr *EndpointManager) WithPeriodicEndpointGC(ctx context.Context, checker EndpointChecker, interval time.Duration) *EndpointManager {
+	mgr.checker = checker
+	controller.NewManager().UpdateController("endpoint-gc",
+		controller.ControllerParams{
+			DoFunc:      mgr.markAndSweep,
+			RunInterval: interval,
+			Context:     ctx,
+		})
+	return mgr
+}
+
+// markAndSweep performs a two-phase garbage collection of endpoints using the
+// configured EndpointChecker.
+//
+// 1) Mark all endpoints that require GC. Do not GC these endpoints this round.
+// 2) Sweep all endpoints marked as requiring GC during the previous iteration.
+//
+// This way, if there is a temporary condition that will be resolved by other
+// components in the system, then we will not flag warnings about the system
+// getting out-of-sync.
+func (mgr *EndpointManager) markAndSweep(ctx context.Context) error {
+	marked := mgr.markEndpoints()
+
+	mgr.mutex.Lock()
+	toSweep := mgr.markedEndpoints
+	mgr.markedEndpoints = marked
+	mgr.mutex.Unlock()
+
+	// TODO: Handle 'ctx' via DeleteEndpoint path
+	return mgr.sweepEndpoints(toSweep)
+}
+
+// markEndpoints runs all endpoints in the manager against the configured
+// EndpointChecker and returns a slice of endpoint ids that require garbage
+// collection.
+func (mgr *EndpointManager) markEndpoints() []uint16 {
+	mgr.mutex.RLock()
+	defer mgr.mutex.RUnlock()
+
+	// TODO: Consider exposing visibility via Endpoint.SetState().
+	needsGC := make([]uint16, 0, len(mgr.endpoints))
+	for eid, ep := range mgr.endpoints {
+		if err := mgr.checker.Check(ep); err != nil {
+			needsGC = append(needsGC, eid)
+		}
+	}
+	return needsGC
+}
+
+// sweepEndpoints iterates through the specified list of endpoints marked for
+// deletion and attempts to garbage-collect them if they still exist.
+func (mgr *EndpointManager) sweepEndpoints(markedEndpoints []uint16) error {
+	toSweep := make([]*endpoint.Endpoint, 0, len(markedEndpoints))
+
+	// 'markedEndpoints' were marked during the previous mark round, so
+	// they may no longer be valid endpoints. Narrow the list to only the
+	// endpoints that remain. Then, release the lock so DeleteEndpoint()
+	// below can independently grab it.
+	mgr.mutex.RLock()
+	for _, id := range markedEndpoints {
+		if ep, ok := mgr.endpoints[id]; ok {
+			toSweep = append(toSweep, ep)
+		}
+	}
+	mgr.mutex.RUnlock()
+
+	errors := 0
+	for _, ep := range toSweep {
+		log.WithFields(logrus.Fields{
+			logfields.EndpointID:  ep.StringID(),
+			logfields.ContainerID: ep.GetShortContainerID(),
+			"k8sPodName":          ep.GetK8sNamespaceAndPodName(),
+			"URL":                 "https://github.com/kubernetes/kubernetes/issues/86944",
+		}).Warning("Stray endpoint found. You may be affected by upstream Kubernetes issue #86944.")
```

Comment:
This obviously points to an upstream k8s issue. This case is not guaranteed to be caused by the upstream issue, but the upstream issue could cause this case. As long as the upstream issue remains open, I think it's reasonable to log this in the Cilium logs. If the linked issue is fixed at some point and this log message still triggers, then I think it's reasonable to re-evaluate this log message.

https://github.com/kubernetes/kubernetes/issues/86944